### PR TITLE
Report low-level Puma errors

### LINKED ISSUE #1247: Support Puma lowlevel_error_handler
You have a puma plugin. This is really great.

https://puma.io/puma/index.html#error-handling

It would be great, if the plugin would register directly a `lowlevel_error_handler`.
Or alternatively, if you could document its usage at https://docs.appsignal.com/ruby/integrations/puma.html

Considerations/learnings from others:

- https://github.com/puma/puma/discussions/3135

### LINKED ISSUE #1269: Error raised in Middlware not caught
I have an Exception, raised from a Middleware, that has not been caught be Appsignal. Another library caught the error correctly.
I created a reproduction repo and gave you access @tombruijn. You need to setup `.env` and the necessary environment variables. You can write to me for any help or in a call so I can show you the behavior if you cannot reproduce it.

---

## Describe the bug

A SystemStackError is raised in a Middleware and Appsignal does not track it

## To Reproduce

Clone the reproduction git repo I shared with you and set it up. It's a Rails App. Set all the env variables (I can help with that if necessary) and load the homepage. An exception will be raised from the middleware and caught by an error tracker but not from Appsignal.

### PR DESCRIPTION
**Pull request by tombruijn**
## Extract Rack request metadata helpers to class

Create a new class for the Rack request metadata extraction so it can be reused in other places more easily too.

Some of the specs are duplicated. I've moved all the specs for edge cases to the class with the actual implementation. The AbstractMiddleware just tests the basics that we expect that class to handle.

## Install Puma hook only for Puma 3 and newer

Set some limit for our Puma hook so we don't break apps using an older Puma. Version 2 and 3 have the biggest difference I could find for the things we rely on so I've set the limit there.

## Report low-level Puma errors

When an error is reported by Puma's `lowlevel_error` reporter, it doesn't arrive at our `EventHandler`'s `on_error` callback or through the `AbstractMiddleware` instrumentation. This resulted in the report in issue #1269 of an error not being reported.

Report the error through a patch on the `Puma::Server` class. I first wanted to add it as described in their docs: https://puma.io/puma/index.html#error-handling This method only works if you can configure it through the Puma config file. It would also require us to determine the response in the block given to that setting. This patch will call the super method, which will determine the response and also allows applications to specify their own `lowlevel_error` handler.

Use the `Appsignal::Rack::ApplyRackRequest` helper to set the request metadata on the transaction, so it's hopefully a bit easier for people who view errors reported this way to debug the issue. This error doesn't flow normally through our instrumentation, so it needs this metadata set manually like this. The transaction is closed through the `RACK_AFTER_REPLY` hook set in the EventHandler. Usuallly, the EventHandler still creates the transaction for the request. If the EventHandler doesn't create the transaction because it's not set up correctly, the `report_error` helper will create a new transaction for this error.

Since version 4 of the Ruby gem we've gotten a couple of reports of applications suddenly reporting new errors, usually internal errors developers can't really fix. This patch will ignore a couple of internal errors from Puma as well to prevent those issues from appearing in the next couple of days.

Closes #1247 Fixes #1269

## Fix Hanami request params not being set

The Hanami integration has a different way from setting the params from other Rack apps. Set it manually, rather than rely on the `ApplyRackRequest` helper.

Add logic to not even try to set the params if the `params_method` option is set to nil.

### CODE REVIEW

File: lib/appsignal/integrations/puma.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            super(error, env)
+          end
+
+        unless PumaServerHelper.ignored_error?(error)
+          Appsignal.report_error(error) do |transaction|
+            Appsignal::Rack::ApplyRackRequest
+              .new(::Rack::Request.new(env))
+              .apply_to(transaction)
+            transaction.add_tags(
+              :reported_by_puma_lowlevel_error => true,
```
Comment (unflxw) [inline lib/appsignal/integrations/puma.rb:21]: Nit: since our global search only allows searching for tag _values_, `:reported_by => :puma_lowlevel_error` might make it easier to search for them.

(I think this tag is a great idea, by the way)

**Review by unflxw — APPROVED** [review]
Do you have an example of the kinds of errors that this will report? I'm wondering if we'd want it to be off-by-default, or toggle-able via a config option.

**Comment by tombruijn** [comment]
> Do you have an example of the kinds of errors that this will report? I'm wondering if we'd want it to be off-by-default, or toggle-able via a config option.

Not sure exactly. The reported error in #1269 was a SystemStackError, but I'm sure there are more. To avoid the noise I already ignored some errors. Let's see if people report any false error incidents.
